### PR TITLE
API: `finch.astype` function

### PR DESCRIPTION
**Pull request by mtsokol**
Hi @willow-ahrens @hameerabbasi!

This PR introduces Array API [`astype` function](https://data-apis.org/array-api/latest/API_specification/generated/array_api.astype.html) (without `device` keyword arg for now).

The test suite will pass once https://github.com/willow-ahrens/Finch.jl/pull/438 is merged and a new release is published.

### CODE REVIEW

**Comment by willow-ahrens** [comment]
Awesome! If we wanted to match https://numpy.org/doc/stable/reference/generated/numpy.ndarray.astype.html, we would need casting and order arguments. I do not understand the order argument for np.astype, so I'll need a little help there, but I think the casting argument just errors depending on what kind of type combination is requested, right?

**Comment by hameerabbasi** [comment]
> Awesome! If we wanted to match https://numpy.org/doc/stable/reference/generated/numpy.ndarray.astype.html, we would need casting and order arguments. I do not understand the order argument for np.astype, so I'll need a little help there, but I think the casting argument just errors depending on what kind of type combination is requested, right?

This page is down right now, but it's really what we should be targeting rather than NumPy: https://data-apis.org/array-api/latest/

**Comment by willow-ahrens** [comment]
do we have a snapshot of the page, or is there a markdown somewhere?

**Comment by hameerabbasi** [comment]
It's based off this repo: https://github.com/data-apis/array-api

I've alerted the people responsible to the outage.

**Review by willow-ahrens — APPROVED** [review]
LGTM!

File: src/finch/dtypes.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    (np.float32, jl.Float32),
+    (np.float64, jl.Float64),
+    (np.complex64, jl.ComplexF32),
+    (np.complex128, jl.ComplexF64),
+    (np.bool_, jl.Bool),
+]
+
+dtype_py_to_jl = {pair[0]: pair[1] for pair in _dtype_pairs}
+
+dtype_jl_to_py = {pair[1]: pair[0] for pair in _dtype_pairs}
```
Comment (hameerabbasi) [inline src/finch/dtypes.py]: We don't really need this -- the Array API standard stipulates that we can use any dtype objects, not necessarily NumPy ones. I think something as simple as

```python
int_ = jl.Int
int8 = jl.Int8
...
```

might suffice.

**Review by hameerabbasi — CHANGES_REQUESTED** [review]
LGTM besides the one comment about extra complexity.

**Comment by hameerabbasi** [comment]
> > Awesome! If we wanted to match https://numpy.org/doc/stable/reference/generated/numpy.ndarray.astype.html, we would need casting and order arguments. I do not understand the order argument for np.astype, so I'll need a little help there, but I think the casting argument just errors depending on what kind of type combination is requested, right?
> 
> This page is down right now, but it's really what we should be targeting rather than NumPy: https://data-apis.org/array-api/latest/

This page is back up, by the way.

**Review by hameerabbasi — APPROVED** [review]

File: src/finch/dtypes.py (hunk 1 above)
Comment (mtsokol) [reply src/finch/dtypes.py]: That's right - done!

**Comment by hameerabbasi** [comment]
This looks great, thanks, @mtsokol!

**Pull request merged**